File: Online2/G/q.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

using ll = long long;
using pii = pair<ll, ll>;

const ll INF = 1e18;

void Merge(vector<pii> &vec)
{
    sort(vec.begin(), vec.end());
    vector<pii> res;
    res.push_back(vec[0]);
    for (size_t i = 1; i < vec.size(); ++i)
    {
        if (vec[i].first <= res.back().second)
        {
            res.back().second = max(res.back().second, vec[i].second);
        }
        else
        {
            res.push_back(vec[i]);
        }
    }
    vec = res;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    int n;
    ll d;
    cin >> n >> d;

    vector<pii> vec(n);
    for (int i = 0; i < n; i++)
    {
        cin >> vec[i].first >> vec[i].second;
    }

    Merge(vec);
    n = vec.size();

    ll ans = 0;
    ll prv = -INF;

    for (int i = 0; i < n; i++)
    {
        vec[i].first = max(vec[i].first, prv + d);
        ll temp = (vec[i].second - vec[i].first + d) / d;
        ans += temp;
        prv = vec[i].first + (temp - 1) * d;
    }

    cout << ans << '\n';
    return 0;
}

File: Online2/G/q.py
import sys

INF = 10**18

n, d = map(int, input().split())
vec = [tuple(map(int, input().split())) for _ in range(n)]

vec.sort()
res = [vec[0]]
for i in range(1, len(vec)):
    if vec[i][0] <= res[-1][1]:
        res[-1] = (res[-1][0], max(res[-1][1], vec[i][1]))
    else:
        res.append(vec[i])
vec = res

ans = 0
prv = -INF

for i in range(len(vec)):
    start = max(vec[i][0], prv + d)
    temp = (vec[i][1] - start + d) // d
    ans += temp
    prv = start + (temp - 1) * d

print(ans)
